File: commons-aws/src/main/java/pl/chilldev/commons/aws/sns/TypedNotificationHandler.java
/**
 * This file is part of the ChillDev-Commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2017 - 2018 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package pl.chilldev.commons.aws.sns;

import java.util.function.Consumer;

import com.fasterxml.jackson.databind.ObjectMapper;
import pl.chilldev.commons.aws.MessageHandler;

/**
 * SNS notifications handler that processes typed message.
 */
public class TypedNotificationHandler extends SimpleNotificationHandler
{
    /**
     * Initializes SNS handler.
     *
     * @param objectMapper JSON handler.
     * @param messageHandler Single message consumer.
     * @param type Message content type.
     * @param <Type> Message type.
     */
    public <Type> TypedNotificationHandler(
        ObjectMapper objectMapper,
        Consumer<Type> messageHandler,
        Class<Type> type
    )
    {
        super(new MessageHandler<>(objectMapper, messageHandler, type)::handle);
    }
}

File: commons-aws/src/main/kotlin/pl/wrzasq/commons/aws/messaging/sns/TypedNotificationHandler.kt
/*
 * This file is part of the pl.wrzasq.commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2017 - 2019, 2021 - 2022 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package pl.wrzasq.commons.aws.messaging.sns

import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json
import pl.wrzasq.commons.aws.messaging.MessageHandler

/**
 * SNS notifications handler that processes typed message.
 *
 * @param json JSON handler.
 * @param messageHandler Single message consumer.
 * @param typeSerializer Expected message type handler.
 * @param <Type> Message type.
 */
class TypedNotificationHandler<Type>(
    json: Json,
    messageHandler: (Type) -> Unit,
    typeSerializer: KSerializer<Type>
) : SimpleNotificationHandler(MessageHandler(json, messageHandler, typeSerializer)::handle)
